sql/core/src/main/scala/org/apache/spark/sql/execution/MetadataOnlyOptimizer.scala

```diff
@@ -0,0 +1,190 @@
+      val partitionColumns = files.partitionSchema.map { field =>
+        attributeMap.getOrElse(field.name, throw new AnalysisException(
+          s"Unable to resolve ${field.name} given [${l.output.map(_.name).mkString(", ")}]"))
+      }
+      val filterColumns = filters.flatMap(_.references)
+      val projectSet = parent.references ++ AttributeSet(filterColumns)
+      if (projectSet.subsetOf(AttributeSet(partitionColumns))) {
+        val selectedPartitions = files.location.listFiles(filters)
+        val partitionValues = selectedPartitions.map(_.values)
+        val valuesRdd = sparkSession.sparkContext.parallelize(partitionValues, 1)
+        val valuesPlan = LogicalRDD(partitionColumns, valuesRdd)(sparkSession)
+        if (projectList.nonEmpty) {
+          parent.withNewChildren(Project(projectList, valuesPlan) :: Nil)
+        } else {
+          parent.withNewChildren(valuesPlan :: Nil)
+        }
+      } else {
+        parent
+      }
+
+    case relation: CatalogRelation =>
+      val attributeMap = relation.output.map(attr => (attr.name, attr)).toMap
+      val partitionColumns = relation.catalogTable.partitionColumnNames.map { column =>
+        attributeMap.getOrElse(column, throw new AnalysisException(
+          s"Unable to resolve ${column} given [${relation.output.map(_.name).mkString(", ")}]"))
+      }
+      val filterColumns = filters.flatMap(_.references)
+      val projectSet = parent.references ++ AttributeSet(filterColumns)
+      if (projectSet.subsetOf(AttributeSet(partitionColumns))) {
+        val partitionColumnDataTypes = partitionColumns.map(_.dataType)
+        val partitionValues = catalog.getPartitionsByFilter(relation.catalogTable, filters)
+          .map { p =>
+            InternalRow.fromSeq(
+              partitionColumns.map(a => p.spec(a.name)).zip(partitionColumnDataTypes).map {
+                case (rawValue, dataType) => Cast(Literal(rawValue), dataType).eval(null)
+              })
+          }
+        val valuesRdd = sparkSession.sparkContext.parallelize(partitionValues, 1)
+        val valuesPlan = LogicalRDD(partitionColumns, valuesRdd)(sparkSession)
+        val filterPlan =
+          filters.reduceLeftOption(And).map(Filter(_, valuesPlan)).getOrElse(valuesPlan)
+        if (projectList.nonEmpty) {
+          parent.withNewChildren(Project(projectList, filterPlan) :: Nil)
+        } else {
+          parent.withNewChildren(filterPlan :: Nil)
+        }
+      } else {
+        parent
+      }
+
+    case _ =>
+      parent
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!sparkSession.sessionState.conf.optimizerMetadataOnly) {
+      return plan
+    }
+    plan.transform {
+      case a @ Aggregate(_, _, child) if canSupportMetadataOnly(a) =>
+        val (plan, projectList, filters, _) = findRelation(child)
+        if (plan.isDefined) {
+          convertToMetadataOnlyPlan(a, projectList, filters, plan.get)
```

Comment:
instead of collecting project list and filter conditions, produce a `LogicalRDD` directly, can we just replace the table relation with `LogicalRDD` and still keep these `Filter` and `Project` operators?